pkg/sql/alter_column_type.go

```diff
@@ -231,14 +222,75 @@ func alterColumnTypeGeneral(
 
 	shadowColName := sqlbase.GenerateUniqueConstraintName(col.Name, nameExists)
 
-	// The default computed expression is casting the column to the new type.
-	newComputedExpr := tree.CastExpr{
-		Expr:       &tree.ColumnItem{ColumnName: tree.Name(col.Name)},
-		Type:       toType,
-		SyntaxMode: tree.CastShort,
+	var newColComputeExpr *string
+	// oldCol still needs to have values written to it in case nodes read it from
+	// it with a TableDescriptor version from before the swap.
+	// To achieve this, we make oldCol a computed column of newCol using
+	// inverseExpr.
+	// If a USING EXPRESSION was provided, we cannot generally invert
+	// the EXPRESSION to insert into the old column and thus will force an
+	// error using the computed expression. Any inserts into the new column
+	// will fail until the old column is dropped.
+	var inverseExpr string
+	if Using != nil {
+		// Ensure the USING EXPRESSION returns the new type.
+		// Replace column references with typed dummies to allow typechecking.
+		replacedExpr, _, err := replaceVars(tableDesc, Using)
```

Comment:
I've seen this logic in a lot of places now -- is there something in the `schemaexpr` package that does this for you?

Reply:
I created a helper function to do this replace and sanitize that seems to be a reoccuring pattern.